### PR TITLE
Improve backchannel logout to be triggered during session_terminate event flow and improve sso session binding to check the commonAuthId validity

### PR DESCRIPTION
**Pull request by piraveena**
Resolves https://github.com/wso2/product-is/issues/10419

Front port the below PRs
https://github.com/wso2-support/identity-inbound-auth-oauth/pull/774
https://github.com/wso2-support/identity-inbound-auth-oauth/pull/780

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -2586,6 +2593,49 @@ private String appendAuthenticatedIDPs(SessionDataCacheEntry sessionDataCacheEnt
...
+     *
+     * @param sessionDataCacheEntry SessionDataCacheEntry.
+     * @param opbscookie            opbscookie value.
+     */
+    private void storeOpbsInSessionContext(SessionDataCacheEntry sessionDataCacheEntry, String opbscookie) {
+
+        String sessionContextIdentifier = getSessionContextIdentifier(sessionDataCacheEntry);
+        if (StringUtils.isNotBlank(sessionContextIdentifier)) {
+            SessionContext sessionContext = FrameworkUtils.getSessionContextFromCache(sessionContextIdentifier);
+            if (sessionContext != null) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2607]: shall we merge the nested if condition with this?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/SSOSessionBasedTokenBinder.java [hunk 2]
```diff
@@ -113,6 +118,28 @@ public void clearTokenBindingElements(HttpServletRequest request, HttpServletRes
...
 
+        try {
+            String sessionIdentifier = getTokenBindingValue((HttpServletRequest) request);
+            if (StringUtils.isBlank(sessionIdentifier)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("CommonAuthId cookie is not found in the request.");
+                }
+                return false;
+            }
+            // Retrieve session context information using sessionIdentifier in order to check the validity of
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/impl/SSOSessionBasedTokenBinder.java:129]: let's use multi-line comments

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/handler/OIDCLogoutEventHandler.java [hunk 3]
```diff
@@ -85,6 +81,73 @@ public String getName() {
...
+
+    private String getopbsCookieId(Event event) {
+
+        HttpServletRequest request = getHttpRequestFromEvent(event);
+        String opbsCookie = null;
+        if (request != null) {
+            // Get the opbscookie from request.
+            opbsCookie = getOpbsCookieFromRequest(request);
+        }
+        if (StringUtils.isEmpty(opbsCookie)) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/handler/OIDCLogoutEventHandler.java:106]: let's use StringUtils.isNotBlank

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 1 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2607]: there are two else part here

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/DefaultLogoutTokenBuilder.java [hunk 4]
```diff
@@ -213,20 +238,17 @@ private String getSigningTenantDomain(OAuthAppDO oAuthAppDO) {
     /**
      * Returns the OIDCsessionState of the obps cookie.
      *
-     * @param request
+     * @param opbscookie
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/DefaultLogoutTokenBuilder.java:241]: Better if we could add a description to this param for consistency.

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutRequestSender.java [hunk 5]
```diff
@@ -89,18 +106,19 @@ public void sendLogoutRequests(HttpServletRequest request) {
         }
     }
 
+
     /**
      * Returns a Map with logout tokens and back-channel logut Url of Service providers.
      *
-     * @param request
+     * @param opbsCookie
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutRequestSender.java:113]: Better if we could add a description to this param for consistency.

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutTokenBuilder.java [hunk 6]
```diff
@@ -38,4 +38,14 @@
      */
     Map<String, String> buildLogoutToken(HttpServletRequest request)
             throws IdentityOAuth2Exception, InvalidOAuthClientException;
+
+    /**
+     * Returns logout token and back-channel logout uri map.
+     *
+     * @param opbsCookie opbscookie value
+     * @return a map of logout tokens and corresponding back-channel logout URLs.
+     * @throws IdentityOAuth2Exception
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/backchannellogout/LogoutTokenBuilder.java:47]: Shall we use sentence-case here for the descriptions?

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/handler/OIDCLogoutEventHandler.java [hunk 7]
```diff
@@ -85,6 +81,73 @@ public String getName() {
...
     }
 
+    private boolean isLogoutInitiatedFromOIDCApp(Event event) {
+
+        HttpServletRequest request = getHttpRequestFromEvent(event);
+        if (request != null) {
+            if (StringUtils.equals(request.getParameter(TYPE), FrameworkConstants.RequestType.CLAIM_TYPE_OIDC)) {
+                // If a logout request is triggered from an OIDC app then the OIDCLogoutServlet
+                // and OIDCLogoutEventHandler both are triggered and the logout request is handled in both
+                // places. https://github.com/wso2/product-is/issues/6418
```
Comment (ChamathNS) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/handler/OIDCLogoutEventHandler.java:91]: Shall we make use of a comment block here?

**Comment by piraveena** [comment]
PR builder is running here https://github.com/wso2/product-is/actions/runs/394203565

**Review by emswbandara — APPROVED** [review]

**Pull request merged**